### ISSUE TITLE
pvr.nextpvr crashing kodi with "ERROR: AddOnLog: NextPVR PVR Client: XXXXX Duration set to 0 XXXXX"

### ISSUE THREAD

**Issue by wmarler** [kodi-pvr/pvr.nextpvr#152]
I think this might be a duplicate bug of https://github.com/kodi-pvr/pvr.nextpvr/issues/145 or https://github.com/xbmc/xbmc/issues/18368 . However, I'm not sure, so filing just in case.

I am using releases https://github.com/kodi-pvr/pvr.nextpvr/releases/tag/3.3.21-Leia of pvr.nextpvr and https://github.com/xbmc/xbmc/releases/tag/18.8-Leia of kodi/xbmc. 

This is a reliably reproducible crash. kodi will crash on Linux (Arch) and my nVidia Shield when skipping forward during a recording. I gathered the following logs from Linux (easier to get v Shield, plus gdb). Note that I do have an EDL file, but that doesn't seem to be an issue (I reproduced without the EDL file at all, and with the EDL file having no contents). 

[kodi_crashlog-20201011_223423.log](https://github.com/kodi-pvr/pvr.nextpvr/files/5362969/kodi_crashlog-20201011_223423.log)

**Comment by wmarler** [comment]
(the .ts file plays fine when loaded directly into kodi)

**Comment by emveepee** [comment]
@wmarler Unless you are trying to report an issue with the code with a possible fix please post the support request on the NextPVR forum using the NextPVR support policy https://github.com/sub3/NextPVR/wiki/get-help Logs are often easiest to send via the Kodi log uploader or using paste.kodi.tv

**Comment by emveepee** [comment]
@wmarler this does look like the same issue you noted.  Installing Matrix ton your Shield you should be able to confirm that the fix worked but if the file still has problem I could give it a further look.  Note that no new updates are planned for Leia.

**Comment by wmarler** [comment]
@emveepee thanks, I'll see if I can test. 

Could the fix merit/qualify for a backport to Leia? Not an "update" per se, but afaict Matrix is still beta, and this is a pretty disruptive issue, maybe worth it? (I certainly understand the need for upgrading, and personally love the "the old version is deprecated and the new version isn't ready yet" mentality, but the other members of my household need stability). 

I wasn't able to see if there was a workaround outside of kodi/pvr.nextpvr. I don't really understand the root cause here (looked like the recorded file not being closed properly?), but could the crashing of kodi be mitigated by running shell commands (e.g. `fmpeg –i corrupted_input.ts –map –ignore_unknown/-copy_unknown –c copy fixed.ts`, or a straight up `cp old.ts new.ts; rm -rf old.ts; mv new.ts old.ts`) or similar that I can script as part of PostProcessing.sh?

**Comment by emveepee** [comment]
> 
> 
> @emveepee thanks, I'll see if I can test.
> 
> Could the fix merit/qualify for a backport to Leia? Not an "update" per se, but afaict Matrix is still beta, and this is a pretty disruptive issue, maybe worth it? (I certainly understand the need for upgrading, and personally love the "the old version is deprecated and the new version isn't ready yet" mentality, but the other members of my household need stability).
> 

Crashing is certainly disruptive but you mentioned one sample, and I only have one sample too.  I asked you to post in the NextPVR forum for a reason you aren't going to see find out if it is a common problem here on github.  I suspect not though, we've never had reports about this problem going back to xbmc days and the change for Matrix wasn't even released until after alpha2.   As I understand things you also didn't even confirm using Matrix, that the new logic prevents the crash.

Basically I don't have information to ask @phunkyfish or others spend time to approve a PR against Leia if you were able to submit one.

> I wasn't able to see if there was a workaround outside of kodi/pvr.nextpvr. I don't really understand the root cause here 

The only thing unusual about the sample file I had is that it was from Merit UK with padding before the station was on the air, then no audio until the full broadcast began and even with the patch it only avoid crashing navigation was still off.  The ts stream was valid.

**Comment by phunkyfish** [comment]
Leia is frozen now. There are no further merges allowed.

Please just confirm if the issue is fixed on Matrix. That is the only way forward for this now

**Comment by wmarler** [comment]
> Crashing is certainly disruptive but you mentioned one sample, and I only have one sample too. I asked you to post in the NextPVR forum for a reason you aren't going to see find out if it is a common problem here on github. 

I just tested on a Windows VM. Playback was fine, no crashes, even skipping commercials via the EDL. So this is a Linux-specific problem.

I can't test on Matrix on the Shield without sideloading. I haven't checked to see if that's even possible, but since you're mentioning it it sounds like it is ... I'm reluctant anyway b/c I have a mysql backend with a media library in addition to the PVR functionality, and upgrading to Matrix creates a new database version, and I'll have some split-brain to deal with (one db for Matrix clients, one db for Leia clients).

> Leia is frozen now. There are no further merges allowed.

It sounds like that's what must be done though ... ok. I'll get back to you.

**Comment by emveepee** [comment]
For Android, if you are running MySsql you likely have figured out how to enter advancesetttings.xml so just rename it before starting the test.  If you read up on adb debugging via IP or USB it can be pretty easy.

**Comment by wmarler** [comment]
It's actually easier for me to build kodi and the pvr.nextpvr from the git repo and test on Arch Linux rather than sideload a package onto the Shield. 

In Matrix, skipping forward doesn't crash Kodi, but it still exits out of recording playback. I turned on debug logging and captured the attached log (I trimmed some of my personal stuff out). 

Specifically I think it's this that causes the crash:

```
2020-10-15 22:04:53.192 T:130907   ERROR <general>: AddOnLog: pvr.nextpvr: Duration set to 0
2020-10-15 22:04:53.192 T:130907   DEBUG <general>: CInputStreamPVRRecording - ClosePVRStream - closed recording stream pvr://recordings/tv/active/NFL Football/NFL%20Football (1601)%20Philadelphia%20Eagles%20at%20Pittsburgh%20Steelers, TV%20(KDVR-DT), 20201011_170542, 2019.pvr
2020-10-15 22:04:53.192 T:130907   DEBUG <general>: SeekTime - seek ended up on time -2147483648
```

In the Leia logs there is a similar message: `2020-10-11 22:34:19.752 T:140527219496512   ERROR: AddOnLog: NextPVR PVR Client: XXXXX Duration set to 0 XXXXX`

[kodi.log](https://github.com/kodi-pvr/pvr.nextpvr/files/5389185/kodi.log)

**Comment by emveepee** [comment]
@wmarler yourt log shows me a a potential addon problem can you make the file available?  I'd like the edl file too.  I know it is large but you mentioned a Windows VM so you probably have OneDrrive

@phunkyfish it also shows a core problem which I think I have explained before.  There is no way for the addon to report an error condition to core that coree accepts.  -1 returns are ignored when PVR_ERORR is not a possibel return code and even returning a PVR_ERROR when it is an option doesn't help.  The addon has to wait unit core gives up.

**Comment by emveepee** [comment]
@wmarler there is something else I would like.  Could you please send me the XML output from so I can get the details on the recording.

http://tv.wmarler.com:8866/service?method=recording.list&recording_id=2019

**Comment by wmarler** [comment]
I'm not certain you need the .ts file as I can reproduce the playback exit with every recording I have on hand. The exit occurs the first time I skip forward, whether that's manual, or automatic via the EDL. But sure, I can share. I've attached the .edl and .xml files here (github didn't like them with their natural file extensions, but was ok with me appending the .txt ). I'll send the link to the Google Drive location of the .ts when it's complete (I only have the free tier of Microsoft One Drive, 10GB, not enough for this recording).

[NFL Football_20201011_11001400.edl.txt](https://github.com/kodi-pvr/pvr.nextpvr/files/5393123/NFL.Football_20201011_11001400.edl.txt)
[NFL Football_20201011_11001400.xml.txt](https://github.com/kodi-pvr/pvr.nextpvr/files/5393124/NFL.Football_20201011_11001400.xml.txt)

**Comment by wmarler** [comment]
the .ts file is here: https://drive.google.com/file/d/14e5-dAUJ0laA9cGQvM0e8iz2zYNVokig/view?usp=sharing
Please let me know when you've pulled it down so I can disable sharing/remove it.

**Comment by emveepee** [comment]
Sorry I didn't realize it was every file so probably not worth sending the big file.  Too late since it is there we might as well use the same thing and I've started.   I don't need the XML file, I need the API  result.

**Comment by wmarler** [comment]
The output of https://tv.wmarler.com:8866/service?method=recording.list&recording_id=2019 is 'An unexpected TLS packet was received':

```
% wget 'https://tv.wmarler.com:8866/service?method=recording.list&recording_id=2019'
--2020-10-16 11:09:45--  https://tv.wmarler.com:8866/service?method=recording.list&recording_id=2019
Resolving tv.wmarler.com (tv.wmarler.com)... 10.10.1.99, 2606:4700:3035::6812:3f24, 2606:4700:3035::ac43:8b28, ...
Connecting to tv.wmarler.com (tv.wmarler.com)|10.10.1.99|:8866... connected.
GnuTLS: An unexpected TLS packet was received.
Unable to establish SSL connection.
```

using http it succeeds:
```
% wget 'http://tv.wmarler.com:8866/service?method=recording.list&recording_id=2019' 
--2020-10-16 11:19:32--  http://tv.wmarler.com:8866/service?method=recording.list&recording_id=2019
Resolving tv.wmarler.com (tv.wmarler.com)... 10.10.1.99, 2606:4700:3033::6812:3e24, 2606:4700:3035::ac43:8b28, ...
Connecting to tv.wmarler.com (tv.wmarler.com)|10.10.1.99|:8866... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1917 (1.9K) [text/xml]
Saving to: ‘service?method=recording.list&recording_id=2019’

service?method=recording. 100%[====================================>]   1.87K  --.-KB/s    in 0s      

2020-10-16 11:19:32 (238 MB/s) - ‘service?method=recording.list&recording_id=2019’ saved [1917/1917]
```

[2019.txt](https://github.com/kodi-pvr/pvr.nextpvr/files/5393288/2019.txt)

**Comment by emveepee** [comment]
Thanks you'd have to go through a few hopes to have setup NextPVR for https.  There is on oddity in there but I have to check some more things.  You can delete the file now.

**Comment by wmarler** [comment]
Awesome, thanks.

**Comment by emveepee** [comment]
I am running on ubunutu right now doing some Android stuff and with linux server and stock Kodi it is fine.

Your log has a -1 in a  read which is the Kodi input stream seek failing.   What backend platform are you using?

```
2020-10-15 22:04:50.510 T:130907   DEBUG <general>: Open - probing detected format [mpegts]
2020-10-15 22:04:50.530 T:130907   DEBUG <general>: AddOnLog: pvr.nextpvr: Seek: Seek:37  0  229376 0 -1
2020-10-15 22:04:50.530 T:130907   ERROR <general>: ffmpeg[0x563512d17360X]: [mpegts] Unable to seek back to the start
```

I would need to see the backend logs to see how it handles this seek.

Another thing to try for your self compile is to comment out this line https://github.com/kodi-pvr/pvr.nextpvr/blob/Matrix/src/Recordings.cpp#L252  It is not clear why but NextPVR is sending a -1 as a resume point.

**Comment by wmarler** [comment]
My backend is Ubuntu 18.04.5, I installed nextpvr via the nextpvr-helper package:

```
% dpkg -l | grep -i next
ii  nextpvr-helper                                     1.4                                              all          nextpvr requirement helper
```
According to the settings.html, I'm running version 5.0.8.200823

I dumped the logs from the NextPVR backend using Settings > Logs Files, but looking through them at the 22:04:50 time they didn't have anything useful (just lots of `Received 17361625160 bytes from device so far`). Looks like things got rotated out. 

So I bounced the nextpvr service, and regenerated the failure. kodi.log snippet:

```
2020-10-16 20:24:52.639 T:5189    DEBUG <general>: Open - probing detected format [mpegts]
2020-10-16 20:24:52.644 T:5189    DEBUG <general>: AddOnLog: pvr.nextpvr: Seek: Seek:37  0  1081344 0 -1
2020-10-16 20:24:52.644 T:5189    ERROR <general>: ffmpeg[0x56489aee5cc0X]: [mpegts] Unable to seek back to the start
```
... then discovered that my client & server are out of sync timewise by almost 2 minutes -_-. On the backend the request came in at 20:26:39

nrecord.log snippet:

```
2020-10-16 20:26:39.223 [DEBUG][6]      Got request [10.10.1.99]: /live
2020-10-16 20:26:39.227 [DEBUG][6]      /live?recording=2019&client=XBMC-6cabc983e0404095ac7ebaf59f4ec8ab
2020-10-16 20:26:39.229 [DEBUG][6]      range: bytes=0-
2020-10-16 20:26:39.245 [DEBUG][6]      StreamNative@1..
2020-10-16 20:26:39.245 [DEBUG][6]      StreamNative@2
2020-10-16 20:26:39.245 [DEBUG][6]      StreamNative@3
2020-10-16 20:26:39.245 [DEBUG][6]      XBMC client sid is: 6cabc983e0404095ac7ebaf59f4ec8ab
2020-10-16 20:26:39.245 [DEBUG][6]      StreamNative@4
2020-10-16 20:26:39.252 [DEBUG][6]      RollingFile.Open(/media/nextpvr/nextpvr-ota-recordings/NFL Football/NFL Football_20201011_11001400.ts) ..
2020-10-16 20:26:39.260 [DEBUG][6]      About to stream recording: /media/nextpvr/nextpvr-ota-recordings/NFL Football/NFL Football_20201011_11001400.ts
2020-10-16 20:26:39.261 [DEBUG][6]      Request had range request
2020-10-16 20:26:39.261 [DEBUG][6]      Requested Range: bytes=0- (23759418192 bytes)
2020-10-16 20:26:39.505 [DEBUG][4]      Got request [10.10.1.99]: /service (recording.edl)
2020-10-16 20:26:39.505 [DEBUG][4]      method=recording.edl
2020-10-16 20:26:39.505 [DEBUG][4]      parameters: 
2020-10-16 20:26:39.505 [DEBUG][4]         method: recording.edl
2020-10-16 20:26:39.505 [DEBUG][4]         recording_id: 2019
2020-10-16 20:26:39.505 [DEBUG][4]         sid: 6cabc983e0404095ac7ebaf59f4ec8ab
2020-10-16 20:26:39.505 [DEBUG][4]         client_ip: 10.10.1.99
2020-10-16 20:26:39.505 [DEBUG][4]         user_agent: Kodi/19.0-ALPHA2 (X11; Linux x86_64) Arch Linux/rolling App_Bitness/64 Version/19.0-ALPHA2-(18.9.702)-Git:20201015-f68475a8d5f
2020-10-16 20:26:39.505 [DEBUG][4]         host_callback: tv.wmarler.com:8866
2020-10-16 20:26:39.520 [DEBUG][4]      GetRecordingEDL
2020-10-16 20:26:39.521 [DEBUG][4]      <?xml version="1.0" encoding="utf-8" ?>
<rsp stat="failed">
  <err code="6" msg="Recording EDL not found" />
</rsp>

2020-10-16 20:26:41.892 [DEBUG][6]      Client has disconnected
2020-10-16 20:26:41.892 [DEBUG][6]      Sent total of 33445200 bytes
```

Full logs:
[nrecord.log](https://github.com/kodi-pvr/pvr.nextpvr/files/5394889/nrecord.log)
[kodi.log](https://github.com/kodi-pvr/pvr.nextpvr/files/5394891/kodi.log)

I'll rebuild with that line commented and see if it fixes anything.

**Comment by wmarler** [comment]
I commented out `tag.SetLastPlayedPosition(XMLUtils::GetIntValue(pRecordingNode, "playback_position"));` but playback still exited. I still have the same `AddOnLog: pvr.nextpvr: Seek: Seek:37  0  1081344 0 -1` in my kodi.log that doesn't seem right.

**Comment by emveepee** [comment]
I didn't think that would be it but sub won't be sending -1 anymore after the next release.   Do you get that -1 as the last variable on small recordings too?   What about if you you install Kodi on your ubuntu backend and test with it?

**Comment by wmarler** [comment]
Define a "small" recording for me please :). Like 30 minutes? 
The Ubuntu backend is headless, but I'll spin up an Ubuntu VM and test in it.

**Comment by emveepee** [comment]
You said any recording so 30 minutes is fine if you already have one but anything short is fine to test if the -1 is file system or size related.

**Comment by emveepee** [comment]
@wmarler does using an IP address impact this?   Perhaps there is something in ffmpeg or the backend that flags the page as being unseekable.  I couldn't duplicate that option with a hosts file though.  

Also did you edit the nrecord.log to change the client IP of the Arch and Shield devices?  Are you testing this through the Internet?

**Comment by wmarler** [comment]
Apologies, I haven't been at the keys all weekend. I am still intending to spin up a Ubuntu VM, 

I am not modifying the log files, nor running through the internet. I am running a pihole DNS inside my network that resolves tv.wmarler.com to 10.10.1.99. How do you think the IP layer could be interfering though? It seems the client and server are communicating (the stream does begin to play, and will play fine as long as I don't attempt to skip forward). 

I updated NextPVR to 5.0.9.201018 and playback still exits.

**Comment by emveepee** [comment]
The logs all show the clients using the same IP I found that unusual just wanted to confirm it was LAN and not timeout related.

**Comment by emveepee** [comment]
@wmarler  if you enable extra curl logging in Kodi I think you will find that the recordings are not returning a content-length.  Try and put your Kodi debug log on a paste site.

**Comment by wmarler** [comment]
Ok. Here's a log grab with curl logging specified (and using a 30-minute recorded show, btw): https://pastebin.com/ndHwLebe .

I see a Content-Length entry at 14:27:32.031: `2020-10-27 14:27:32.031 T:1430296   DEBUG <general>: Curl::Debug - HEADER_IN: Content-Length: 107`

However this comes after `2020-10-27 14:27:31.520 T:1430456   DEBUG <general>: AddOnLog: pvr.nextpvr: Seek: Seek:37  71940  10551296 0 -1`

Maybe a race condition?

**Comment by emveepee** [comment]
I have asked @sub3 to  look at the server.  I did ask you to turn off compression but it still look like it is on in that log.  

Also that log shows new information since not only are running DNS you are running a squid proxy server.  Did you ever try without a proxy server?

**Comment by wmarler** [comment]
BOOM. PROXY SERVER.
When the proxy server is disabled, no crash. Playback works with Matrix, in my Arch setup.
on the Shield, still on Leia, I get a spinner, but something looks wonky. I'll give that some more work later tonight.

**Comment by wmarler** [comment]
Also: when you wrote "turn off compression" I'm sorry, I missed that. What compression are you referring to?

**Comment by emveepee** [comment]
@wmarler I apologize I asked another users having the same problem to do it.  In config.xml change this to false <UseCompression>true</UseCompression> and restart the server.   From this it does sound like the proxy service is optimizing the headers because clearly the content-length is being send.  Maybe that is on compression since the calls with content-length I specifically turn off compression in the header  

Without this header Kodi won't allow seeking it doesn't get from the content-range bytes.

**Comment by emveepee** [comment]
I can duplicate this with squid so I think this can be closed as an issue with the add-on or Kodi.

**Comment by emveepee** [comment]
@wmarler here is a Matrix patch that I tested that uses an older API call that works for me with a proxy https://paste.kodi.tv/dokuvoceqa The web server responds differently with this call.

**Comment by emveepee** [comment]
@wmarler this issue can be closed.  @sub3  has updated the backend and the next version will have new response headers that will send the content-length to Kodi through squid.  Your issue was never client related, but I have used the information here to prevent skipping (except resume) on unseekable file in a new PR and not just handle the resulting errors.

**Comment by wmarler** [comment]
thank you @emveepee & @sub3 !
